monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/MergePrioritizedListObservable.scala

```diff
@@ -0,0 +1,157 @@
+
+    // MUST BE synchronized by `lock`
+    var lastAck = Continue: Future[Ack]
+
+    case class PQElem(data: A, promise: Promise[Ack], priority: Int) extends Comparable[PQElem] {
+      override def compareTo(o: PQElem): Int =
+        priority.compareTo(o.priority)
+    }
+
+    val pq = new PriorityBlockingQueue[PQElem](sources.size)
+
+    // MUST BE synchronized by `lock`
+    var completedCount = 0
+
+    // MUST BE synchronized by `lock`
+    def rawOnNext(a: A): Future[Ack] = {
+      if (isDone) Stop else out.onNext(a)
+    }
+
+    def processNext(): Future[Ack] = {
+      val e = pq.remove()
+      val fut = rawOnNext(e.data)
+      e.promise.completeWith(fut)
+      fut
+    }
+
+    // MUST BE synchronized by `lock`
+    def signalOnNext(): Future[Ack] =
+      lock.synchronized {
+        lastAck = lastAck match {
+          case Continue => processNext()
+          case Stop => Stop
+          case async =>
+            async.flatMap {
+              // async execution, we have to re-sync
+              case Continue => lock.synchronized(processNext())
+              case Stop => Stop
+            }
+        }
+
+        lastAck
+      }
+
+    def signalOnError(ex: Throwable): Unit =
+      lock.synchronized {
+        if (!isDone) {
+          isDone = true
+          out.onError(ex)
+          lastAck = Stop
+          completePromises()
+        }
+      }
+
+    def signalOnComplete(): Unit =
+      lock.synchronized {
+        completedCount += 1
+
+        if (completedCount == numberOfObservables && !isDone) {
+          lastAck match {
+            case Continue =>
+              isDone = true
+              out.onComplete()
+              completePromises()
+            case Stop =>
+              () // do nothing
+            case async =>
+              async.onComplete {
+                case Success(Continue) =>
+                  lock.synchronized {
+                    if (!isDone) {
+                      isDone = true
+                      out.onComplete()
+                      completePromises()
+                    }
+                  }
+                case _ =>
+                  () // do nothing
+              }
+          }
+
+          lastAck = Stop
+        }
+      }
+
+    def completePromises(): Unit = {
+      pq.iterator().asScala.foreach(e => e.promise.complete(Success(Stop)))
+    }
+
+    val composite = CompositeCancelable()
+
+    sources.zip(priorities).foreach { pair =>
+      composite += pair._1.unsafeSubscribeFn(new Subscriber[A] {
+        implicit val scheduler: Scheduler = out.scheduler
+
+        def onNext(elem: A): Future[Ack] = {
+          if (isDone) {
+            Stop
+          } else {
+            pq.add(PQElem(elem, Promise(), pair._2))
+            signalOnNext()
+          }
+        }
+
+        def onError(ex: Throwable): Unit =
+          signalOnError(ex)
+
+        def onComplete(): Unit =
+          signalOnComplete()
```

Comment:
I think there's a race here where you can get an `onComplete` before the value passed in the final `onNext` has actually processed.

Reply:
Can you elaborate?

Reply:
1) `onNext` gets called with the final value for this observable, but this observable isn't the highest priority and there's a value waiting, so the `signalOnNext` call triggers the sending of a different value
1) upstream is now exhausted, and isn't obliged to wait for the ack, so it triggers `onComplete`, which calls `signalOnComplete`
1) `onComplete` then increments `completedCount` when there is still an outstanding value for that observable not yet sent
1) If this happens to be the last incomplete observable, then we call `out.onComplete` while there is still at least one value outstanding

Reply:
Will look into this when I work on this again. I hope to be able to this week.

Reply:
Fixed by adding back `lock.synchronized` to the `onNext` method (I'd dropped that in copying over the code from `CombinedLatestListObservable`). Since `onNext` will be called before `onComplete`, the latter will now not trigger downstream `onComplete` until the last downstream `onNext` completes.

Reply:
I'm not sure that actually fixes the scenario I'm describing?

Reply:
I'm not sure that actually fixes issue in the situation I've described?

I think the easiest way to make sure that completion doesn't get ahead of the last value is to store the promise created in `onNext`, and then `def onComplete() = lastFuture.foreach(_ => signalOnComplete())`

Reply:
The key is that so long as there are items the downstream hasn't yet processed, `lastAck` will remain a `Future[Ack]`, and `signalOnComplete` won't call `out.onComplete` until that future completes, i.e., downstream has finished processing all pending items including the last one from the last source.

Take the simplest case of your example: 2 sources w/ 1 item each:
* Source A calls `onNext`, item gets sent downstream: `lastAck = Future[Ack for A's item]`
* Source A calls `onComplete`: `completedCount = 1`
* Source B calls `onNext`, item gets queued: `lastAck = Future[Ack for A's item].flatMap(Future[Ack for B's item])`
* Source B calls `onComplete`: `completedCount = 2; lastAck.onComplete { ... out.onComplete }`